packages/block/block.ts

```diff
@@ -0,0 +1,216 @@
+/* eslint-disable @typescript-eslint/unbound-method */
+import {
+  childNodes$,
+  cloneNode$,
+  createEventListener,
+  insertBefore$,
+  insertText,
+  innerHTML$,
+  remove$ as removeElement$,
+  setAttribute,
+  setText,
+  mapGet$,
+  mapHas$,
+  mapSet$,
+} from './dom';
+import { renderToTemplate } from './template';
+import { AbstractBlock, EditType, Hole } from './types';
+import type { Edit, EditChild, Props, VElement } from './types';
+
+export const createBlock = (fn: (props?: Props) => VElement) => {
+  let cachedProp: string;
+  let cachedHole: Hole;
+  const vnode = fn(
+    new Proxy(
+      {},
+      {
+        // A universal getter will return a Hole instance if props[any] is accessed
+        // Allows code to identify holes in virtual nodes ("digs" them out)
+        get(_, prop: string): Hole {
+          if (prop === cachedProp) return cachedHole;
+          const hole = new Hole(prop);
+          cachedProp = prop;
+          cachedHole = hole;
+          return hole;
+        },
+      },
+    ),
+  );
```

Comment:
@aidenybai IMHO, this could be an implementation error here.

Here I have an example:

```tsx
const block = createBlock(({ text, className }) => <div className={className}>{text}</div>);

block({ text: 'Million.js is the Best!', className: 'text-white' });
block({ text: 'Million.js is the Fastest!', className: 'text-black' });
```

Let's break it up:

```
- text: 'Million.js is the Best!'

cachedProp -> text
cachedHole -> new Hole<'text'>

- className: 'text-white'

cachedProp<'text'> !== 'className'
cachedHole -> new Hole<'className'>

- text: 'Million.js is the Fastest!'

cachedProp<'className'> !== 'text'
cachedHole -> new Hole<'text'>
```

And now, the `Hole` instance is being created constantly, causing a huge performance issue.

Reply:
My original idea is that we could hold a cache for every block:

```ts
export const createBlock = <P = {}>(fn: ((prop: P) => VElement)) => {
  const cache = new Map<keyof P, P[keyof P]>();
  const vnode = fn(new Proxy(
    {},
    {
      // A universal getter will return a Hole instance if props[any] is accessed
      // Allows code to identify holes in virtual nodes ("digs" them out)
      get(_, prop: string): Hole {
        if (cache.has(prop)) return cache.get(prop);
        const hole = new Hole(prop);
        cache.set(prop, hole);
        return hole;
      },
    },
  ));
}
```

Reply:
My original idea about `lastProp` was that: the shape of `prop` could have changed, so we might want to re-create every hole instance when the prop shape changes.

But it seems that I was wrong, the shape of the prop is not going to change. So let's throw away the `lastProp` here.

Reply:
Makes sense, seems like a very reasonable approach + thanks for the helpful examples! Pushing a change now.

Reply:
![image](https://user-images.githubusercontent.com/38025074/219585335-bc30da48-b300-4dc8-802b-040d454aca2c.png)

Reply:
Have you set up a benchmark suite for new Block API yet? I'd like to try it locally to see if I can spot anything.

Reply:
Old benchmarks will be deprecated in favor of JS Framework Benchmark